File: content/snippets/scala2/interview_lcci_16_14_best_line.scala
object Lcci1614BestLine {
    private def collinear(a: Array[Int], b: Array[Int], c: Array[Int]): Boolean = {
        val x1 = b(0).toLong - a(0).toLong
        val y1 = b(1).toLong - a(1).toLong
        val x2 = c(0).toLong - a(0).toLong
        val y2 = c(1).toLong - a(1).toLong
        x1 * y2 == x2 * y1
    }

    def bestLine(points: Array[Array[Int]]): Array[Int] = {
        val n = points.length
        var bestI = 0
        var bestJ = 1
        var bestCount = 2
        for (i <- 0 until n) {
            for (j <- i + 1 until n) {
                var count = 2
                for (k <- j + 1 until n) {
                    if (collinear(points(i), points(j), points(k))) count += 1
                }
                if (count > bestCount || (count == bestCount && (i < bestI || (i == bestI && j < bestJ)))) {
                    bestI = i
                    bestJ = j
                    bestCount = count
                }
            }
        }
        Array(bestI, bestJ)
    }

    def main(args: Array[String]): Unit = {
        println(bestLine(Array(Array(0, 0), Array(1, 1), Array(1, 0), Array(2, 2))).mkString("[", ", ", "]"))
    }
}

File: content/snippets/python/interview_lcci_16_14_best_line.py
from typing import List


def collinear(a: List[int], b: List[int], c: List[int]) -> bool:
    x1, y1 = b[0] - a[0], b[1] - a[1]
    x2, y2 = c[0] - a[0], c[1] - a[1]
    return x1 * y2 == x2 * y1


def best_line(points: List[List[int]]) -> List[int]:
    n = len(points)
    best_i, best_j, best_count = 0, 1, 2
    for i in range(n):
        for j in range(i + 1, n):
            count = 2
            for k in range(j + 1, n):
                if collinear(points[i], points[j], points[k]):
                    count += 1
            if count > best_count or (count == best_count and (i < best_i or (i == best_i and j < best_j))):
                best_i, best_j, best_count = i, j, count
    return [best_i, best_j]


def main() -> None:
    print(best_line([[0, 0], [1, 1], [1, 0], [2, 2]]))


if __name__ == "__main__":
    main()
